File: 0_100/010_Regular_Expression_Matching/Solution.rs
impl Solution {
    pub fn is_match(s: String, p: String) -> bool {
        let mut memo = HashMap::new();
        Solution::dp(s.as_bytes(), 0, p.as_bytes(), 0)
    }

    fn dp(s: &[u8], i: usize, p: &[u8], mut j: usize) -> bool {
        if j == p.len() {
            return i == s.len();
        }
        if i == s.len() {
            if (p.len() - j) % 2 == 1 {
                return false;
            }
            while j + 1 < p.len() {
                if p[j + 1] != '*' as u8 {
                    return false;
                }
                j += 2
            }
            return true;
        }
        if s[i] == p[j] || p[j] == '.' as u8 {
            if j < p.len() - 1 && p[j + 1] == '*' as u8 {
                Solution::dp(s, i, p, j + 2) || Solution::dp(s, i + 1, p, j)
            } else {
                Solution::dp(s, i + 1, p, j + 1)
            }
        } else {
            if j < p.len() - 1 && p[j + 1] == '*' as u8 {
                Solution::dp(s, i, p, j + 2)
            } else {
                false
            }
        }
    }
}

File: 0_100/010_Regular_Expression_Matching/Solution.py
from functools import lru_cache


class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        self.memo = {}
        return self.dp(s, 0, p, 0)

    @lru_cache(maxsize=None)
    def dp(self, s, i, p, j):
        if j == len(p):
            return i == len(s)
        if i == len(s):
            if (len(p) - j) % 2 == 1:
                return False
            while j+1 < len(p):
                if p[j+1] != '*':
                    return False
                j += 2
            return True
        res = False
        if s[i] == p[j] or p[j] == '.':
            if j < len(p) - 1 and p[j+1] == '*':
                res = self.dp(s, i+1, p, j) or self.dp(s, i, p, j+2)
            else:
                res = self.dp(s, i+1, p, j+1)
        else:
            if j < len(p) - 1 and p[j+1] == '*':
                res = self.dp(s, i, p, j+2)
            else:
                res = False
        return res
